### PR TITLE
Add support for EXTCODEHASH

### PR DESCRIPTION
**Pull request by jwasinger**
Implementation for `EXTCODEHASH` opcode support in Solidity.  For now, this seems to work with LLL assembly, which is what I am using to write some `EXTCODEHASH` state tests.

Implements https://eips.ethereum.org/EIPS/eip-1052

### CODE REVIEW

File: libevmasm/Instruction.h [hunk 1]
```diff
@@ -82,6 +82,7 @@ enum class Instruction: uint8_t
 	EXTCODECOPY,		///< copy external code (from another contract)
 	RETURNDATASIZE = 0x3d,	///< get size of return data buffer
 	RETURNDATACOPY = 0x3e,	///< copy return data in current environment to memory
+	EXTCODEHASH = 0x3f,
```
Comment (axic) [inline libevmasm/Instruction.h]: Needs description.

File: libevmasm/Instruction.cpp [hunk 2]
```diff
@@ -215,6 +216,7 @@ static const std::map<Instruction, InstructionInfo> c_instructionInfo =
 	{ Instruction::EXTCODECOPY,	{ "EXTCODECOPY",	0, 4, 0, true, Tier::ExtCode } },
 	{ Instruction::RETURNDATASIZE,	{"RETURNDATASIZE",	0, 0, 1, false, Tier::Base } },
 	{ Instruction::RETURNDATACOPY,	{"RETURNDATACOPY",	0, 3, 0, true, Tier::VeryLow } },
+	{ Instruction::EXTCODESIZE,	{ "EXTCODEHASH",	0, 1, 1, false, Tier::ExtCode } },
```
Comment (axic) [inline libevmasm/Instruction.cpp]: This seems to be wrong?

File: libevmasm/GasMeter.cpp [hunk 3]
```diff
@@ -112,6 +112,7 @@ GasMeter::GasConsumption GasMeter::estimateMax(AssemblyItem const& _item, bool _
 			gas += wordGas(GasCosts::copyGas, m_state->relativeStackElement(-2));
 			break;
 		case Instruction::EXTCODESIZE:
+		case Instruction::EXTCODEHASH:
 			gas = GasCosts::extCodeGas(m_evmVersion);
```
Comment (axic) [inline libevmasm/GasMeter.cpp:115]: Is this really charged the same?

**Comment by axic** [comment]
Needs a changelog entry and tests too (see tests for create2).

**Comment by chriseth** [comment]
Will take this over.

File: libevmasm/GasMeter.cpp (hunk 3 above)
Comment (chriseth) [reply libevmasm/GasMeter.cpp:115]: Most likely.

Comment (chriseth) [reply libevmasm/GasMeter.cpp:115]: Ah I'm sorry, - it seems to be the same as `Instruction::BALANCE`.

**Comment by chriseth** [comment]
Is it accepted for constantinople in the same way as shift instructions are?

**Comment by axic** [comment]
@chriseth rebase and fixed this.

**Review by axic — DISMISSED** [review]

**Comment by axic** [comment]
@chriseth I think this is fine to merge. (Also needed for creating tests in github.com/ethereum/tests)

**Comment by axic** [comment]
@chriseth added documentation

File: docs/assembly.rst [hunk 4]
```diff
@@ -271,12 +271,14 @@ In the grammar, opcodes are represented as pre-defined identifiers.
...
+| extcodehash(a)          |     | C | code hash of address a                                          |
++-------------------------+-----+---+-----------------------------------------------------------------+
 | create(v, p, s)         |     | F | create new contract with code mem[p...(p+s)) and send v wei     |
 |                         |     |   | and return the new address                                      |
 +-------------------------+-----+---+-----------------------------------------------------------------+
 | create2(v, n, p, s)     |     | C | create new contract with code mem[p...(p+s)) at address         |
-|                         |     |   | keccak256(<address> . n . keccak256(mem[p...(p+s))) and send v  |
-|                         |     |   | wei and return the new address                                  |
+|                         |     |   | keccak256(0xff . <address> . n . keccak256(mem[p...(p+s)))      |
+|                         |     |   | and send v wei and return the new address                       |
```
Comment (axic) [inline docs/assembly.rst:281]: This was missed in #5013.

File: docs/assembly.rst [hunk 5]
```diff
@@ -271,12 +271,14 @@ In the grammar, opcodes are represented as pre-defined identifiers.
...
 +-------------------------+-----+---+-----------------------------------------------------------------+
+| extcodehash(a)          |     | C | code hash of address a                                          |
++-------------------------+-----+---+-----------------------------------------------------------------+
 | create(v, p, s)         |     | F | create new contract with code mem[p...(p+s)) and send v wei     |
 |                         |     |   | and return the new address                                      |
 +-------------------------+-----+---+-----------------------------------------------------------------+
 | create2(v, n, p, s)     |     | C | create new contract with code mem[p...(p+s)) at address         |
-|                         |     |   | keccak256(<address> . n . keccak256(mem[p...(p+s))) and send v  |
-|                         |     |   | wei and return the new address                                  |
+|                         |     |   | keccak256(0xff . <address> . n . keccak256(mem[p...(p+s)))      |
```
Comment (chriseth) [inline docs/assembly.rst:280]: could use `u256(n)` to clarify padding?

**Review by chriseth — APPROVED** [review]

File: docs/assembly.rst (hunk 5 above)
Comment (axic) [reply docs/assembly.rst:280]: Started doing that but ran into a lot more to explain. Will prepare a separate PR.

**Pull request merged**